File: 20-actors/fuchi/methods/allocate.py
"""allocate.py — 扶持 (fuchi) maintainer sustenance allocation. ADR-2606052300.

THE HEART of the actor and the charter-clean inverse of an investment fund's cap-table.

A VC fund computes equity stakes and a return waterfall. 扶持 computes, for a cohort of
covenant-bound maintainers (信者) who keep etzhayyim's actors alive, each one's:

  - weight        w = ln(1 + min(tenure_years, 40)) * hazard   (the Displacement-Dividend
                                                                 curve, ADR-2606032130)
  - share         w_i / Σ w_j over the VOWED cohort            (sums to 1; PRIORITY, not cash)
  - priority_rank 1 = provisioned first under a scarce stage-capacity cap
  - floor         in-kind sustenance floor at elapsed=0, stage-capped, decaying over 5y

The single invariant that makes this charter-clean and NOT an investment vehicle:

  * cash_usd_micros is structurally 0 for every allocation (cash≡0, ADR-2605301020 N1);
  * the instrument is one of {in-kind-grant, sustenance, tooling-access, compute-access} —
    equity / debt / convertible / revenue-share / profit-claim / carry / dividend are a
    ValueError, never an allocation (G1, Charter-Rider §2(b));
  * the maintainer's work product is commons (owns_payoff is structurally False, G5).

There is no NAV, no IRR, no exit, no liquidation. A "share" governs only the SEQUENCING and
the in-kind floor — exactly what allocate.py in 50-infra/etzhayyim-public-fund/displacement
does for displaced workers, reused here for active maintainers.

Stdlib only. Deterministic.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field

TENURE_CAP_YEARS = 40.0
HAZARD_MIN = 1.0
HAZARD_MAX = 2.0
HORIZON_YEARS = 5.0

# G1 — sustenance instruments only (mirror of the ontology :alloc/instrument :db/allowed).
ALLOWED_INSTRUMENTS = ("in-kind-grant", "sustenance", "tooling-access", "compute-access")
# The investment-fund vocabulary that is UNREPRESENTABLE here (defensive denylist; the
# allowlist above is the real gate — this only sharpens the error message).
FORBIDDEN_INSTRUMENTS = (
    "equity", "debt", "convertible", "revenue-share", "profit-claim",
    "carry", "dividend", "loan", "interest", "warrant", "option", "exit",
)


def assert_instrument(instrument: str) -> str:
    """G1 INVARIANT — only a sustenance instrument is allocatable. Anything resembling an
    investment / debt / return claim is a ValueError (not an investment fund)."""
    instr = str(instrument or "").lstrip(":").lower()
    if instr in FORBIDDEN_INSTRUMENTS:
        raise ValueError(
            f"G1: instrument {instr!r} is an investment/return vehicle — UNREPRESENTABLE "
            "(扶持 is sustenance, not a fund; Charter-Rider §2(b))"
        )
    if instr not in ALLOWED_INSTRUMENTS:
        raise ValueError(f"G1: instrument {instr!r} not in {ALLOWED_INSTRUMENTS}")
    return instr


@dataclass(frozen=True)
class Maintainer:
    did: str
    tenure_months: int            # 勤続 months of mission service
    hazard_permille: int          # [1000, 2000] -> 1.0 .. 2.0 toil-hazard
    maintains: tuple[str, ...] = ()      # actor handles kept alive
    prior_imputed_usd_micros_yr: int = 0  # in-kind valuation only; NEVER cash
    covenant: str = "vowed"       # "outreach" (minimal floor) | "vowed" (full sustenance)
    owns_payoff: bool = False     # G5 — structurally False; work product is commons


@dataclass(frozen=True)
class Allocation:
    maintainer_did: str
    instrument: str               # G1 — one of ALLOWED_INSTRUMENTS
    weight: float
    share: float                  # fraction of cohort PRIORITY; Σ == 1 over vowed cohort
    priority_rank: int            # 1 = provisioned first under a scarce cap
    floor_usd_micros_yr: int      # in-kind floor at elapsed=0, stage-capped
    cash_usd_micros: int = 0      # INVARIANT: always 0 (cash≡0, N1)
    server_held_key: bool = False  # G9 — always False (no-server-key)

    def __post_init__(self) -> None:
        # The structural proofs, asserted at construction (defence in depth alongside schema).
        if self.cash_usd_micros != 0:
            raise ValueError("cash≡0 INVARIANT (G2/N4): 扶持 never disburses cash")
        if self.server_held_key:
            raise ValueError("no-server-key INVARIANT (G9): allocation is member/Council-signed")
        assert_instrument(self.instrument)


def _capped_tenure_years(tenure_months: int) -> float:
    return min(tenure_months / 12.0, TENURE_CAP_YEARS)


def _hazard(hazard_permille: int) -> float:
    h = hazard_permille / 1000.0
    if not (HAZARD_MIN <= h <= HAZARD_MAX):
        raise ValueError(f"hazard out of [1.0,2.0]: {h}")
    return h


def tenure_weight(m: Maintainer) -> float:
    """w = ln(1 + min(tenure_years, cap)) * hazard. Log compresses the gradient so a 40y
    maintainer is ~2x a 5y one (not 8x) — honours service without a per-person income
    leaderboard (ADR-2605261000 N6; same curve as Displacement Dividend)."""
    return math.log1p(_capped_tenure_years(m.tenure_months)) * _hazard(m.hazard_permille)


def floor_decay(elapsed_months: int) -> float:
    """decay(t) = clamp(1 - t/HORIZON, 0, 1). The sustenance floor tapers over 5 years as
    the maintainer ascends the Liberation Ladder toward full Basic High Income."""
    t = elapsed_months / 12.0
    return max(0.0, min(1.0, 1.0 - t / HORIZON_YEARS))


def allocate(
    cohort: list[Maintainer],
    stage_ceiling_usd_micros_yr: int,
    elapsed_months: int = 0,
    instrument: str = "sustenance",
) -> list[Allocation]:
    """Allocate tenure-weighted in-kind sustenance over a maintainer cohort.

    Only `vowed` maintainers join the tenure-weighted share pool (the covenant gate, G4).
    `outreach` maintainers receive a minimal floor (share 0) until they vow — they are not
    abandoned, but the full tenure-weighted sustenance is covenant-bound.

    Returns Allocations whose `cash_usd_micros` is structurally 0. Raises if any maintainer
    claims `owns_payoff` (G5) or if `instrument` is an investment vehicle (G1).
    """
    instr = assert_instrument(instrument)
    if any(m.owns_payoff for m in cohort):
        raise ValueError("G5: a maintainer cannot own the payoff — work product is commons")

    vowed = [m for m in cohort if m.covenant == "vowed"]
    total_w = sum(tenure_weight(m) for m in vowed)
    decay = floor_decay(elapsed_months)
    ranked = sorted(vowed, key=tenure_weight, reverse=True)
    rank_of = {m.did: i + 1 for i, m in enumerate(ranked)}

    out: list[Allocation] = []
    for m in cohort:
        if m.covenant == "vowed":
            w = tenure_weight(m)
            share = (w / total_w) if total_w > 0 else 0.0
            rank = rank_of[m.did]
            floor = min(m.prior_imputed_usd_micros_yr, stage_ceiling_usd_micros_yr)
            floor = int(round(floor * decay))
        else:  # outreach — minimal floor, no share (pre-vow)
            w = 0.0
            share = 0.0
            rank = len(vowed) + 1
            floor = int(round(min(m.prior_imputed_usd_micros_yr,
                                  stage_ceiling_usd_micros_yr) * decay * 0.25))
        out.append(Allocation(
            maintainer_did=m.did,
            instrument=instr,
            weight=round(w, 6),
            share=round(share, 6),
            priority_rank=rank,
            floor_usd_micros_yr=floor,
            cash_usd_micros=0,
            server_held_key=False,
        ))
    # vowed allocations first (priority order), then outreach
    out.sort(key=lambda a: (a.priority_rank, -a.weight))
    return out


def cohort_from_seed(records: list[dict]) -> list[Maintainer]:
    """Build a cohort from seed :maintainer/* maps (edn keyword-keyed)."""
    def kw(v):
        return str(v or "").lstrip(":").split("/")[-1].lower()
    out = []
    for r in records:
        out.append(Maintainer(
            did=r.get(":maintainer/did", "?"),
            tenure_months=int(r.get(":maintainer/tenure-months", 0)),
            hazard_permille=int(r.get(":maintainer/hazard-permille", 1000)),
            maintains=tuple(r.get(":maintainer/maintains", []) or []),
            prior_imputed_usd_micros_yr=int(r.get(":maintainer/prior-imputed-usd-micros-yr", 0)),
            covenant=kw(r.get(":maintainer/covenant", ":vowed")),
            owns_payoff=bool(r.get(":maintainer/owns-payoff", False)),
        ))
    return out

File: 20-actors/fuchi/methods/allocate.clj
;; allocate.clj — 扶持 (fuchi) maintainer sustenance allocation.
;;
;; Clojure port of allocate.py (ADR-2606052300), Wave 1 of the clj-native migration
;; (ADR-2606142300) — the FIRST fuchi clj method. THE HEART of the actor and the charter-clean
;; inverse of an investment fund's cap-table: for a cohort of covenant-bound maintainers (信者)
;; who keep etzhayyim's actors alive, computes each one's tenure weight (the Displacement-Dividend
;; curve), share of cohort PRIORITY (sums to 1, NOT cash), priority rank under a scarce stage cap,
;; and a decaying in-kind sustenance floor.
;;
;; The invariants that make this charter-clean and NOT an investment vehicle:
;;   * cash≡0 structurally for every allocation (ADR-2605301020 N1);
;;   * the instrument ∈ {in-kind-grant sustenance tooling-access compute-access} — equity / debt /
;;     convertible / revenue-share / carry / dividend … RAISE (G1, Charter-Rider §2(b));
;;   * the maintainer's work product is commons (owns-payoff structurally false, G5).
;; No NAV, IRR, exit, or liquidation. A "share" governs only SEQUENCING + the in-kind floor.
;;
;; Float math (Math/log1p) is byte-equivalent with allocate.py; round(x,6)/int(round(x)) mirror
;; Python's half-to-even. stdlib only, deterministic.
(ns fuchi.methods.allocate
  (:require [clojure.string :as str])
  (:import [java.math BigDecimal RoundingMode]))

(def tenure-cap-years 40.0)
(def hazard-min 1.0)
(def hazard-max 2.0)
(def horizon-years 5.0)

(def allowed-instruments #{"in-kind-grant" "sustenance" "tooling-access" "compute-access"})  ; G1
(def forbidden-instruments
  #{"equity" "debt" "convertible" "revenue-share" "profit-claim" "carry" "dividend"
    "loan" "interest" "warrant" "option" "exit"})

(defn assert-instrument
  "G1 INVARIANT — only a sustenance instrument is allocatable. Anything resembling an
   investment / debt / return claim RAISES (扶持 is sustenance, not a fund)."
  [instrument]
  (let [instr (str/lower-case (str/replace (str (or instrument "")) #"^:+" ""))]
    (when (forbidden-instruments instr)
      (throw (ex-info (str "G1: instrument " (pr-str instr) " is an investment/return vehicle — "
                           "UNREPRESENTABLE (扶持 is sustenance, not a fund; Charter-Rider §2(b))") {})))
    (when-not (allowed-instruments instr)
      (throw (ex-info (str "G1: instrument " (pr-str instr) " not in " allowed-instruments) {})))
    instr))

(defn- round-n [x scale]
  (.doubleValue (.setScale (BigDecimal/valueOf (double x)) (int scale) RoundingMode/HALF_EVEN)))

(defn- round-int [x]  ; Python int(round(x)) — half-to-even, then integer
  (.longValue (.setScale (BigDecimal/valueOf (double x)) 0 RoundingMode/HALF_EVEN)))

(defn- capped-tenure-years [tenure-months] (min (/ tenure-months 12.0) tenure-cap-years))

(defn- hazard [hazard-permille]
  (let [h (/ hazard-permille 1000.0)]
    (when-not (<= hazard-min h hazard-max) (throw (ex-info (str "hazard out of [1.0,2.0]: " h) {})))
    h))

(defn tenure-weight
  "w = ln(1 + min(tenure-years, cap)) × hazard. Log compresses the gradient so a 40y maintainer
   is ~2× a 5y one (not 8×) — honours service without a per-person income leaderboard."
  [m]
  (* (Math/log1p (capped-tenure-years (:tenure-months m))) (hazard (:hazard-permille m))))

(defn floor-decay
  "decay(t) = clamp(1 − t/HORIZON, 0, 1). The sustenance floor tapers over 5 years toward BHI."
  [elapsed-months]
  (let [t (/ elapsed-months 12.0)]
    (max 0.0 (min 1.0 (- 1.0 (/ t horizon-years))))))

(defn make-allocation
  "Construct an allocation, asserting the structural proofs (cash≡0, no-server-key, instrument)."
  [{:keys [cash-usd-micros server-held-key instrument] :or {cash-usd-micros 0 server-held-key false} :as a}]
  (when (not= 0 cash-usd-micros)
    (throw (ex-info "cash≡0 INVARIANT (G2/N4): 扶持 never disburses cash" {})))
  (when server-held-key
    (throw (ex-info "no-server-key INVARIANT (G9): allocation is member/Council-signed" {})))
  (assert-instrument instrument)
  (merge {:cash-usd-micros 0 :server-held-key false} a))

(defn allocate
  "Allocate tenure-weighted in-kind sustenance over a maintainer cohort. Only `vowed` maintainers
   join the tenure-weighted share pool (covenant gate, G4); `outreach` maintainers receive a
   minimal floor (share 0) until they vow. Raises on owns-payoff (G5) or an investment instrument (G1)."
  ([cohort stage-ceiling] (allocate cohort stage-ceiling 0 "sustenance"))
  ([cohort stage-ceiling elapsed-months instrument]
   (let [instr (assert-instrument instrument)]
     (when (some :owns-payoff cohort)
       (throw (ex-info "G5: a maintainer cannot own the payoff — work product is commons" {})))
     (let [vowed   (filter #(= "vowed" (:covenant %)) cohort)
           total-w (reduce + 0.0 (map tenure-weight vowed))
           decay   (floor-decay elapsed-months)
           ranked  (sort-by tenure-weight (fn [a b] (compare b a)) vowed)  ; desc, stable
           rank-of (into {} (map-indexed (fn [i m] [(:did m) (inc i)]) ranked))
           allocs  (mapv
                    (fn [m]
                      (if (= "vowed" (:covenant m))
                        (let [w (tenure-weight m)]
                          (make-allocation
                           {:maintainer-did (:did m) :instrument instr
                            :weight (round-n w 6)
                            :share (round-n (if (pos? total-w) (/ w total-w) 0.0) 6)
                            :priority-rank (rank-of (:did m))
                            :floor-usd-micros-yr (round-int (* (min (:prior-imputed-usd-micros-yr m 0) stage-ceiling) decay))
                            :cash-usd-micros 0 :server-held-key false}))
                        (make-allocation
                         {:maintainer-did (:did m) :instrument instr
                          :weight 0.0 :share 0.0
                          :priority-rank (inc (count vowed))
                          :floor-usd-micros-yr (round-int (* (min (:prior-imputed-usd-micros-yr m 0) stage-ceiling) decay 0.25))
                          :cash-usd-micros 0 :server-held-key false})))
                    cohort)]
       ;; vowed allocations first (priority order), then outreach
       (vec (sort-by (juxt :priority-rank #(- (:weight %))) allocs))))))

(defn cohort-from-seed
  "Build a cohort from seed :maintainer/* maps (edn keyword-keyed)."
  [records]
  (let [kw (fn [v] (-> (str (or v "")) (str/replace #"^:+" "") (str/split #"/") last str/lower-case))]
    (mapv (fn [r]
            {:did                         (get r :maintainer/did "?")
             :tenure-months               (long (get r :maintainer/tenure-months 0))
             :hazard-permille             (long (get r :maintainer/hazard-permille 1000))
             :maintains                   (vec (get r :maintainer/maintains []))
             :prior-imputed-usd-micros-yr (long (get r :maintainer/prior-imputed-usd-micros-yr 0))
             :covenant                    (kw (get r :maintainer/covenant ":vowed"))
             :owns-payoff                 (boolean (get r :maintainer/owns-payoff false))})
          records)))
